File: LinearSystem/src/lagrange.cpp
#include "Lagrange.h"

Lagrange::Lagrange(int n1) : Interpolation(n1) {}

double Lagrange::solve(double xp)
{
    if(rows == 0)
        throw "No data points";

    double yp = 0;

    for(int i = 0; i < rows; i++)
    {
        double L = 1;

        for(int j = 0; j < rows; j++)
        {
            if(i != j)
            {
                if((*this)(i,0) == (*this)(j,0))
                    throw "duplicate x values";

                L *= (xp - (*this)(j,0)) /
                     ((*this)(i,0) - (*this)(j,0));
            }
        }

        yp += L * (*this)(i,1);
    }

    return yp;
}

File: LinearSystem in Python/src/Lagrange.py
from Interpolation import Interpolation


class Lagrange(Interpolation):
    def __init__(self, n1=0):
        super().__init__(n1)

    def solve(self, xp):
        if self.rows == 0:
            raise Exception("No data points")

        yp = 0.0

        for i in range(self.rows):
            L = 1.0

            for j in range(self.rows):
                if i != j:
                    if self.a[i][0] == self.a[j][0]:
                        raise Exception("duplicate x values")

                    L *= (xp - self.a[j][0]) / (self.a[i][0] - self.a[j][0])

            yp += L * self.a[i][1]

        return yp
